client/state/data-layer/wpcom-http/utils.js

```diff
@@ -42,6 +49,125 @@ export const getHeaders = action => get( action, 'meta.dataLayer.headers', null
  */
 export const getProgress = action => get( action, 'meta.dataLayer.progress', null );
 
+const getRequestStatus = action => {
+	if ( getError( action ) ) {
+		return 'failure';
+	}
+
+	if ( getData( action ) ) {
+		return 'success';
+	}
+
+	return 'pending';
+};
+
+export const getActionKey = fullAction => {
+	const { meta, ...action } = fullAction; // eslint-disable-line no-unused-vars
+
+	return deterministicStringify( action );
+};
+
+/**
+ * Tracks the state of network activity for a given request type
+ *
+ * When we issue _REQUEST type actions they usually create some
+ * associated network activity by means of an HTTP request.
+ * We may want to know what the status of those requests are, if
+ * they have completed or if they have failed.
+ *
+ * This tracker stores the meta data for those requests which
+ * can then be independently polled by React components which
+ * need to know about those data requests.
+ *
+ * Note that this is meta data about remote data requests and
+ * _not_ about network activity, which is why this is code is
+ * here operating on the _REQUEST actions and not in the HTTP
+ * pipeline as a processor on HTTP_REQUEST actions.
+ *
+ * @param {Map} requests stores request meta data; must be Map-like with set/get
+ * @param {Map} requestIds tracks requests by unique ids
+ * @returns {Function} middleware function to track requests
+ */
+export const trackRequests = ( requests, requestIds ) => next => ( store, action ) => {
+	// progress events don't affect
+	// any tracking meta at the moment
+	if ( getProgress( action ) ) {
+		return next( store, action );
+	}
+
+	const actionKey = getActionKey( action );
+	const status = getRequestStatus( action );
+	const meta = requests.get( actionKey ) || {};
+	const requestId = get( action, 'meta.dataLayer.requestId' ) || uniqueId( 'data-request-' );
+
+	const nextMeta = Object.assign(
+		{},
+		meta,
+		{
+			requestId,
+			status,
+		},
+		status !== 'pending' && { lastUpdated: Date.now() }
+	);
+
+	// update the meta
+	requests.set( actionKey, nextMeta );
```

Comment:
Should we set a limit on how many requests we track? A hard limit or an LRU cache maybe?